### PR TITLE
Feature/plmc 267 add complexity parameters to extrinsic benchmarks

### PR DESCRIPTION
**Pull request by JuaniRios**
## What?
Add complexity parameters to the existing extrinsic benchmarks, as well as split them into different benchmarks for each logic path

## Why?
Get a more accurate weight to avoid the chain stalling, and avoid overcharging our users.

## How?
For the complexity parameters, I checked which storage items were being used, and if the amount of reads/writes depended on any 
inputs or previous conditions. For example if an extrinsic needed to get all contributions made by a user, then it meant it had to do
one storage read per existing contribution by that user. So a complexity parameter in this case would be the amount of existing 
contributions for that user before calling the benchmark.

For the logic paths I looked at any if statements.

## Testing?
- Run benchmark tests by calling `cargo test --features runtime-benchmarks bench_`
- Run real benchmarks to see the weight differences by calling `cargo run --features runtime-benchmarks,fast-gov -p polimec-parachain-node -- benchmark pallet --chain=polimec-rococo-local --steps=50 --repeat=20 --pallet=pallet_funding --extrinsic="*" --output="dbg-weights.rs"`
  (WARNING: it will take a long time)

## Anything Else?
- We are still missing the benchmarks for the hrmp extrinsics. But next I will focus on the do_ functions which are called in other places except extrinsics (like on_initialize).

### CODE REVIEW

**Review by lrazovic — CHANGES_REQUESTED** [review]
> The second most difficult thing after writing FRAME benchmarks correctly is to review them. 
> 
> Confucius, 499 BCE

From my analysis and the construction of the worst-case scenarios, it appears we have achieved the goal. The comments received are primarily stylistic, without affecting the benchmarks' overall logic. A review of the guidelines at [Polkadot Blockchain Academy](https://polkadot-blockchain-academy.github.io/pba-book/frame/benchmarking-1/page.html) (and Benchmarking-2) confirms that we are in compliance with all the recommendations.

GG!

File: pallets/funding/src/functions.rs [hunk 1]
```diff
@@ -845,7 +938,16 @@ impl<T: Config> Pallet<T> {
 		// * Emit events *
 		Self::deposit_event(Event::FundsBonded { project_id, amount: plmc_bond, bonder: evaluator.clone() });
 
-		Ok(())
+		let existing_evaluations_count = caller_existing_evaluations.len() as u32;
+		let actual_weight = if existing_evaluations_count == 0 {
```
Comment (lrazovic) [inline pallets/funding/src/functions.rs:942]:
```suggestion
  let actual_weight = if existing_evaluations_count.is_zero()
```

File: pallets/funding/src/functions.rs [hunk 2]
```diff
@@ -867,12 +969,18 @@ impl<T: Config> Pallet<T> {
...
-	) -> DispatchResult {
+	) -> DispatchResultWithPostInfo {
 		// * Get variables *
 		let project_details = ProjectsDetails::<T>::get(project_id).ok_or(Error::<T>::ProjectDetailsNotFound)?;
 		let project_metadata = ProjectsMetadata::<T>::get(project_id).ok_or(Error::<T>::ProjectNotFound)?;
 		let plmc_usd_price = T::PriceProvider::get_price(PLMC_STATEMINT_ID).ok_or(Error::<T>::PriceNotFound)?;
 		let ct_deposit = T::ContributionTokenCurrency::deposit_required(project_id);
+		let existing_bids = Bids::<T>::iter_prefix_values((project_id, bidder)).collect::<Vec<_>>();
+
+		// benchmark variables
```
[the remark this answers is not in the document]
Comment (JuaniRios) [reply pallets/funding/src/functions.rs:980]: Comment is misleading, should be "weight calculation variables". We use them to return to the user the correct weight used, so they are needed always

File: pallets/funding/src/functions.rs [hunk 3]
```diff
@@ -1037,6 +1137,30 @@ impl<T: Config> Pallet<T> {
 		let project_metadata = ProjectsMetadata::<T>::get(project_id).ok_or(Error::<T>::ProjectNotFound)?;
 		let mut project_details = ProjectsDetails::<T>::get(project_id).ok_or(Error::<T>::ProjectDetailsNotFound)?;
 		let ct_deposit = T::ContributionTokenCurrency::deposit_required(project_id);
+		let caller_existing_contributions =
+			Contributions::<T>::iter_prefix_values((project_id, contributor)).collect::<Vec<_>>();
+
+		// Weight flags
+		enum WeightContributionFlag {
```
[the remark this answers is not in the document]
Comment (JuaniRios) [reply pallets/funding/src/functions.rs:1144]: Should we?
It's not used anywhere else, and the enum is just an easier way of representing the data than a long variable name with comments.
I vote keep it in, unless there is some sort of convention I'm not aware of

File: pallets/funding/src/lib.rs [hunk 4]
```diff
@@ -436,6 +440,8 @@ pub mod pallet {
 		type RequiredMaxCapacity: Get<u32>;
 		/// max_message_size config required for the channel from polimec to the project
 		type RequiredMaxMessageSize: Get<u32>;
+		/// max iterations for trying to insert a project on the projects_to_update storage
+		type MaxProjectsToUpdateInsertionAttempts: Get<u32>;
```
[the remark this answers is not in the document]
Comment (JuaniRios) [reply pallets/funding/src/lib.rs:461]: e40634f4

File: pallets/funding/src/lib.rs [hunk 5]
```diff
@@ -955,51 +965,64 @@ pub mod pallet {
...
+			Self::do_start_evaluation(issuer, project_id)
 		}
 
 		/// Starts the auction round for a project. From the next block forward, any professional or
 		/// institutional user can set bids for a token_amount/token_price pair.
 		/// Any bids from this point until the candle_auction starts, will be considered as valid.
 		#[pallet::call_index(3)]
-		#[pallet::weight(WeightInfoOf::<T>::start_auction())]
-		pub fn start_auction(origin: OriginFor<T>, project_id: ProjectId) -> DispatchResult {
+		#[pallet::weight(WeightInfoOf::<T>::start_auction_manually(<T as Config>::MaxProjectsToUpdateInsertionAttempts::get() - 1, 10_000u32))]
```
[the remark this answers is not in the document]
Comment (JuaniRios) [reply pallets/funding/src/lib.rs:995]: Its already explained in line 1041. Should we explain the same parameter every time it's used?

File: pallets/funding/src/functions.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (JuaniRios) [reply pallets/funding/src/functions.rs]: 837e19c6247f17ca048c088f738368b4a15b8d08

File: pallets/funding/src/functions.rs
[the remark this answers is not in the document]
Comment (JuaniRios) [reply pallets/funding/src/functions.rs:1]: done

File: pallets/funding/src/benchmarking.rs
[the remark this answers is not in the document]
Comment (JuaniRios) [reply pallets/funding/src/benchmarking.rs:1]: done

**Review by vstam1 — APPROVED** [review]
As far as I understand, the logic looks good.

**Review by lrazovic — APPROVED** [review]

**Pull request merged**